### PR TITLE
[SDXL] Add SDXL pipeline to SHARK

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit adds SDXL pipeline to SHARK.

Signed-off-by: Abhishek Varma <abhishek@nod-labs.com>

### CODE REVIEW

**Comment by monorimet** [comment]
Cool, can't wait to try this. Let me run it quick and review

File: apps/stable_diffusion/scripts/txt2img.py [hunk 1]
```diff
@@ -16,31 +16,62 @@ def main():
...
+            custom_vae=args.custom_vae,
+            low_cpu_mem_usage=args.low_cpu_mem_usage,
+            debug=args.import_debug if args.import_mlir else False,
+            use_lora=args.use_lora,
+            use_quantize=args.use_quantize,
+            ondemand=args.ondemand,
+        )
+    else:
+        assert (
+            args.height <= 768 and args.width <= 768
```
Comment (monorimet) [inline apps/stable_diffusion/scripts/txt2img.py:54]: I have an issue with splitting these up here based on the specified size.

The first thing people will do with an SDXL model is probably going to be: download SDXL model (or SDXL custom model) -> click on it in the UI dropdown -> enter prompt -> generate image

in this case they will get an error as it's set to 512x512 by default and so it just tries to use the `Text2ImagePipeline` instead of the `Text2ImageSDXLPipeline`.

Can we use the huggingface/civitAI model metadata to determine whether it is an SDXL-based model and then programmatically change the available sizes if yes? This technically works if you're aware of all the boxes to tick and dials to set, but if it breaks the first time you try it, that's bad for us.

I'm open to other points of view on this, but it could be as easy as getting the .ckpt filename and if it has XL we set some `is_sdxl` to True (False by default) and it takes the right path.


A considerable alternative to this is to move XL features to a separate `XL (Beta)` tab and then we won't have to worry about the funky stuff in the SD1.5/2.1 workflow that I ran into with this patch.

File: apps/stable_diffusion/src/pipelines/pipeline_shark_stable_diffusion_utils.py [hunk 2]
```diff
@@ -306,6 +507,99 @@ def produce_img_latents(
...
+            step_start_time = time.time()
+            timestep = torch.tensor([t]).to(dtype).detach().numpy()
+            # expand the latents if we are doing classifier free guidance
+            latent_model_input = torch.cat([latents] * 2)
+
+            latent_model_input = self.scheduler.scale_model_input(
+                latent_model_input, t
+            ).to(dtype)
+
+            # input_to_unet = {
```
Comment (monorimet) [inline apps/stable_diffusion/src/pipelines/pipeline_shark_stable_diffusion_utils.py:536]: Dead code? if this is debug tooling we want to keep let's flag guard it

File: apps/stable_diffusion/web/ui/txt2img_ui.py [hunk 3]
```diff
@@ -156,26 +157,56 @@ def txt2img_inf(
...
+                low_cpu_mem_usage=args.low_cpu_mem_usage,
+                debug=args.import_debug if args.import_mlir else False,
+                use_lora=args.use_lora,
+                use_quantize=args.use_quantize,
+                ondemand=args.ondemand,
+            )
+        else:
+            assert (
+                height <= 768 and width <= 768
+            ), "height/width not in supported range"
```
Comment (monorimet) [inline apps/stable_diffusion/web/ui/txt2img_ui.py:189]: (Refer to above comments on handling XL - ness of a given user config)

**Review by monorimet — CHANGES_REQUESTED** [review]
Tried this out and had a few notes. PTAL

Also, we will eventually need this on top of @aviator19941 's SD implementation in Turbine. The implementation may be slightly cleaner that way. Either way I've suggested a few changes that can make this a little more user-friendly.

Try it out yourself with a custom model, switch between XL models and SD1.5 models, stress test it a bit... and  put cushions on the rough edges and this should be ready to go.

Happy to chat and set up meetings about my comments and/or the impending Turbine integration

(FYI @aviator19941 )

**Comment by Abhishek-Varma** [comment]
Hi @monorimet 

As discussed yesterday, I have addressed the following in the latest push to this revision :-
1. Added SDXL to a separate script in order to not fiddle with current normal SD.
2. Added another tab for SDXL with the set of configuration supported by SHARK SDXL.
3. Got rid of few nit changes.

It is now ready to be re-reviewed.

For CLI : `python apps/stable_diffusion/scripts/txt2img_sdxl.py --prompt="Batman playing electric guitar" --device=rocm --hf_model_id="stabilityai/stable-diffusion-xl-base-1.0"`

**Comment by monorimet** [comment]
Awesome. Thank you @Abhishek-Varma. I'll give it a try later today.

**Comment by monorimet** [comment]
Cool, seems to work. What needs done to have support for custom SDXL models/loras? We can merge once lint is fixed

**Review by monorimet — DISMISSED** [review]

File: apps/stable_diffusion/scripts/txt2img.py [hunk 4]
```diff
@@ -85,4 +85,4 @@ def main():
 
 
 if __name__ == "__main__":
-    main()
+    main()
```
Comment (monorimet) [inline apps/stable_diffusion/scripts/txt2img.py:88]: Unnecessary change maybe

File: apps/stable_diffusion/web/ui/txt2img_ui.py [hunk 5]
```diff
@@ -675,4 +675,4 @@ def txt2img_inf(
                 lora_hf_id,
                 custom_vae,
             ],
-        )
+        )
```
Comment (monorimet) [inline apps/stable_diffusion/web/ui/txt2img_ui.py:678]: nit - indent change? might be fixed by black formatting

**Review by monorimet — CHANGES_REQUESTED** [review]

File: apps/stable_diffusion/src/models/model_wrappers.py [hunk 6]
```diff
@@ -67,6 +73,70 @@ def check_compilation(model, model_name):
...
+    print(f"Saved {module_name} vmfb at {path}")
+    module.load_module(path)
+    return module
+
+
+def process_vmfb_ir_sdxl(extended_model_name, model_name, device, precision):
+    name_split = extended_model_name.split("_")
+    if "vae" in model_name:
+        name_split[5] = "fp32"
+    extended_model_name_for_vmfb = "_".join(name_split)
```
Comment (monorimet) [inline apps/stable_diffusion/src/models/model_wrappers.py:106]: This currently appends the entire device URI to the end of the filename, resulting in filenames like "unet_1_77_1024_1024_fp16_stable-diffusion-xl-base-1_vulkan_00000000_0c00_0000_0000_000000000000.vmfb"

Can we make this conform to the naming conventions for non-SDXL models (without the device URI)? e.g.
"unet_1_64_512_512_fp16_stable-diffusion-2-1-base_vulkan.vmfb"

Comment (Abhishek-Varma) [reply apps/stable_diffusion/src/models/model_wrappers.py:106]: The naming of the vmfbs can be done like that but it will be wrong (and the non-SDXL models should also be updated to have the same) - a vmfb compiled for vulkan AMD device should be differentiated from vulkan NVIDIA device. Something similar is being done for Llama as well : [vicuna.py from Top-Of-Main branch](https://github.com/nod-ai/SHARK/blob/main/apps/language_models/scripts/vicuna.py#L1332-L1343), so instead of the entire URI I'm now using the `args.iree-vulkan-target-triple` flag to add the distinction.

So, now you'll see the filenames like : `clip_1_77_1024_1024_fp16_stable-diffusion-xl-base-1_vulkan_rdna2_unknown_linux.vmfb`

**Comment by Abhishek-Varma** [comment]
> Cool, seems to work. What needs done to have support for custom SDXL models/loras? We can merge once lint is fixed

Hi @monorimet - I've addressed your review comments.

Regarding the custom SDXL/Loras - someone implementing that can follow the same approach used for non-SDXL models with the assumption that the HF repo contains all the sub-models and the base model being used is [stabilityai/stable-diffusion-xl-base-1.0](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0) (since that's what we use in `base_model.json`).

File: apps/stable_diffusion/src/models/model_wrappers.py [hunk 7]
```diff
@@ -86,15 +156,22 @@ def __init__(
...
-        self.height = height // 8
-        self.width = width // 8
+        self.is_sdxl = is_sdxl
+        self.height = height
+        self.width = width
+        if is_sdxl:
+            # We need to scale down the height/width by vae_scale_factor, which
+            # happens to be 8 in this case.
+            self.height = height // 8
+            self.width = width // 8
```
Comment (monorimet) [inline apps/stable_diffusion/src/models/model_wrappers.py:174]: This seems unintentional. The original SD needs `height // 8` and `width //  8`  and this changes the default while assigning the old behavior to SDXL.

If I try SD1.4 from scratch on this commit, I get a 4TB CPU allocation error. If I add `not` to `if is_sdxl` on line 170, the problem seems fixed for SD1.5/2.1 but now the VAE issue starts to make more sense...

I'll update after a bit more debugging.

Comment (monorimet) [reply apps/stable_diffusion/src/models/model_wrappers.py:174]: Oh, yeah I think it might be a good idea to just revert this one change entirely.... unless i'm mistaken

Comment (Abhishek-Varma) [reply apps/stable_diffusion/src/models/model_wrappers.py:174]: Perhaps following would help clear why `// 8` (`vae_scale_factor`) needs to be done for SDXL :-

1. User only mentions the final image size to be generated, be it SD1.X/2.1/XL.
2. In case of SDXL we currently only deal with 1024x1024 image generation.
3. But the inputs to UNET/VAE for a 1024x1024 image needs to be 128x128. And this is nothing but `1024 // 8`.
4. This is something I observed in the normal Python execution by printing the input sizes expected by individual models right before their "forward" method is invoked.
5. Since the inputs to each model is worked out currently via `base_model.json` and for SD1.X/2.1 it substitutes the final height/width - but for SDXL it needs to be the scaled down version (as seen in point 3 above).
6. So, I added that and guarded it with `if is_sdxl` flag.

Comment (monorimet) [reply apps/stable_diffusion/src/models/model_wrappers.py:174]: Before this commit, the SD1.X / 2.1 models already do `height // 8` and `width // 8` here. This commit changes the behavior for those models at this point to `height` and `width` (introduces the CPUAllocator issue) and then defines a special case for XL with the old default behavior. Unless we are reworking the pipeline for SD1.X/2.1 let's keep the non-SDXL path as it was beforehand, to reduce breakages

**Review by monorimet — APPROVED** [review]
